app/api/qstash/workers/sync-users/route.ts

```diff
@@ -0,0 +1,119 @@
+// This is a public API endpoint that will be invoked by QStash.
+import { verifySignatureAppRouter } from "@upstash/qstash/dist/nextjs";
+import { NextRequest, NextResponse } from "next/server";
+import { publishToQstash } from "../../../../../lib/qstash";
+import { UsersSyncPayload } from "../../../../../lib/qstash/types";
+import prisma from "../../../../../lib/prisma";
+import {
+  fetchFarcasterProfiles,
+  UserProfile,
+} from "../../../../../lib/airstack";
+
+async function handler(request: NextRequest) {
+  // Get the payload from the request and extract the offset
+  const body: UsersSyncPayload = await request.json();
+  const { offset } = body;
+
+  // Declaring the batch size
+  const batchSize = 400;
+
+  try {
+    // Read users from the WaitlistedUser Model in the database, grouped by fid with and taken with an offset
+    const usersBatch = await prisma.waitlistedUser.groupBy({
+      by: ["fid"],
+      skip: offset,
+      take: batchSize,
+      orderBy: { fid: "asc" },
+    });
+
+    // Do the syncing work if there is at least one user in the batch (nobody is left behind!)
+    if (usersBatch.length > 0) {
+      // Transforming the fids number array to a string array to satisfy the Airstack query
+      const fidsString = usersBatch.map((fid) => fid.toString());
+
+      // Call Airstack to fetch all useful information about the users in the batch
+      const usersToUpdate: UserProfile[] = [];
+      let pointer = "";
+
+      do {
+        const res = await fetchFarcasterProfiles(fidsString, pointer);
+        if (res) {
+          const { profiles, pageInfo } = res;
+          usersToUpdate.push(...profiles);
+          pointer = pageInfo.nextCursor;
+        }
+      } while (pointer);
+
+      // Creating chunks of users to update in the database in a parallel way
+      const usersToUpdateChunks: UserProfile[][] = [];
+      const chunkSize = 200;
+
+      for (let i = 0; i < usersToUpdate.length; i += chunkSize) {
+        usersToUpdateChunks.push(usersToUpdate.slice(i, i + chunkSize));
+      }
+
+      // Update the users in the database with the new information, in parallel
+      await Promise.all(
```

Comment:
We should process the chunks sequentially, meaning that we iterate on them through a for loop, while then it's correct to keep the Promise.all for the update operations

Reply:
Do you think it would be a problem if the chunks are parallelized as well?

Reply:
Yes, because in that case there's no point in doing chunks as all the calls would still happen in parallel, while it's better to reduce the parallelized queries in 200 by 200

Reply:
Ok, great, I will change it asap 💪🏻